### DISCUSSION TITLE
WAL mode errors

### DISCUSSION THREAD

**Discussion by MaxiStefan** [simolus3/drift#2673] [Q&A] [answered]
Hi,

I tried to see if any other people experienced this sort of issue with WAL mode but I found 0 issues/discussion topics opened on it.

I get the following error whenever I use
db.execute('PRAGMA journal_mode = WAL;'); in the setup callback of NativeDatabase CTOR
![image](https://github.com/simolus3/drift/assets/48346391/6a1920a7-80d4-4bcb-b691-7ced005c7bac)

This is the method I use to create a db connection in an isolate
```
NativeDatabase _openIsolateConnection(String filePath) {
  final file = File(filePath);
  return NativeDatabase(
    file,
    logStatements: true,
    setup: (db) {
      db.execute('PRAGMA busy_timeout = 100;');
    },
  );
}
```

The main thread DB is created using the following method. I have the problem with .createInBackground and without :
```
LazyDatabase _openConnectionWithUserDb({required String fileName}) {
  // the LazyDatabase util lets us find the right location for the file async.
  return LazyDatabase(() async {
    material.WidgetsFlutterBinding.ensureInitialized();

    // put the database file, called db.sqlite here, into the documents folder
    // for your app.
    Directory documentsDirectory = await getApplicationDocumentsDirectory();
    String path = p.join(documentsDirectory.path, fileName);
    File dbFile = File(path);

    if (!(await dbFile.exists())) {
      ByteData data = await rootBundle.load(
        p.join(
          "assets/db",
          "clean_db.sqlite",
        ),
      );
      List<int> bytes =
          data.buffer.asUint8List(data.offsetInBytes, data.lengthInBytes);
      await dbFile.writeAsBytes(bytes);
    }

    return NativeDatabase.createInBackground(
      dbFile,
      logStatements: true,
      setup: (db) {
        // db.execute('PRAGMA journal_mode = WAL;');
        db.execute('PRAGMA busy_timeout = 100;');
      },
    );
  });
}
```

It looks like whenever I open the DB for some batch insert/update on an isolate. I get the error in the Screenshot? Any ideas why?

Additionally, the .wal and .shm databases never seem to go away.

**Comment by MaxiStefan** [comment]
Important detail.... I want to add WAL mode because I've seen in crashlytics that some users run into database is locked errors

**Comment by simolus3** [comment]
> It looks like whenever I open the DB for some batch insert/update on an isolate. I get the error in the Screenshot? Any ideas why?

It looks like the original cause is the "disk IO error" mentioned in the last line which causes transactions to fail, the error you're seeing happens because drift tries to explicitly revert the transactions too.
Does this happen on every (or more than one) device you've tested? The IO error looks suspicious, do you see an earlier statement failing in the transaction?

> Additionally, the .wal and .shm databases never seem to go away.

They should be deleted once the last connection to the database closes. If you have a database that you're not closing, it can be that the files continue to exist.

Are these isolates created for multiple short-lived tasks? If so, are you closing databases on isolates? I'm using finalizers in `package:sqlite3` that will try to close databases if they're no longer referenced in Dart, but those probably won't fire for a closing isolate.

As a workaround until we figure this out, you could try to [share an existing database connection](https://drift.simonbinder.eu/docs/advanced-features/isolates/#simple-sharing) with spawned isolate. This will lead to only one sqlite3 connection, preventing "database is locked" errors. This only works for isolates created in Dart though, not for things like a background work manager job.

**Reply by MaxiStefan** [reply]
@simolus3  Thanks for the quick reply!

> Does this happen on every (or more than one) device you've tested? The IO error looks suspicious, do you see an earlier statement failing in the transaction?

I've seen this on a Samsung S8+ (Android 9), One+ Nord T2(Android 13), Samsung Galaxy Active Tab (Android 12). No earlier statement failing. A batch upsert computation just suddenly ends into that DISK I/O error

For the isolates:
1) I use the compute method to create an isolate to perform some database related tasks
2) I use the _openIsolateConnection method to open a new connection on the isolate and at the end of the isolate I run db.close();

I've recently attempted to use the .createInBackground CTOR of the NativeDatabase seeing as it manages an isolate for me, however, I started having state management problems with certain states not being there when I expect them due to db operations (perhaps?) taking a bit longer. Which is understandable given that data needs to be passed from the isolate to my main thread. 


I am considering moving forward with either ._**createInBackground**_ (and do the necessary structural changes throughout the app) or ._**computeWithDatabase**_ for sharing. Which one would you consider a better approach?

P.S. I tried to play around with computeWithDatabase in the recent past and I failed to pass the data fetched from the database inside the computation callback to a variable that calls the function (e.g., final x = await fetchBulkData()). Is that at all possible or is the computation isolated?

**Reply by simolus3** [reply]
> Which one would you consider a better approach?

I think the two approaches work well together, actually. `createInBackground` makes the database class in your main isolate use a background isolate to run the queries (but SQL is still built in the main isolate, and results are mapped in the main isolate as well).

`computeWithDatabase` makes an existing database connection available to a short-lived task. This is useful when you have things like bulk inserts that parse a large response from a server (which can be expensive on its own). I think `computeWithDatabase` without `createInBackground` can create problems as well: If you then create a isolate needing the database, all of the database operation that isolate makes are executed on the main isolate.

I'd use `createInBackground` generally, and for compute-heavy tasks that need to be on a background isolate either way _and also_ need the database, `computeWithDatabase` is a good approach.

> Is that at all possible or is the computation isolated?

The return value from the callback should be transferred to the calling isolate, yes. Something like this should be possible:

```dart
Future<List<Row>> loadBulkData(MyDatabase db) async {
  await db.computeWithDatabase(
    connect: MyDatabase.new,
    computation: (db) async {
      final data = await fetchRowsFromNetwork();
      await db.batch((batch) {
        // More expensive work like inserting data
      });

      return data;
    },
  );
}
```

**Reply by MaxiStefan** [reply]
Right, I am going to play around in an isolated environment to see how these hold up and possibly be back if I have further question. Thank you very much for the time you spent on this ! :)

**Comment by MaxiStefan** [comment]
So i've completely migrated every isolate that I want to offload work to to .computeWithDatabase and I have use .createInBackround db by default. Everything works pretty well, however, if I activate the WAL mode on a freshly created database, the DISK I/O error comes back, if I disable it, it's fine.


Given this:

> As a workaround until we figure this out, you could try to [share an existing database connection](https://drift.simonbinder.eu/docs/advanced-features/isolates/#simple-sharing) with spawned isolate. This will lead to only one sqlite3 connection, preventing "database is locked" errors. This only works for isolates created in Dart though, not for things like a background work manager job.

Should I give up on trying to use WAL mode? Since it would be redundant?

**Reply by MaxiStefan** [reply]
Not sure if this is of any help but this is the error that I get. Coincidently, the error only occurs for this dao/table. If i remove it while WAL is active, it's gone. IF I keep the operation without WAL it's gone

**Reply by MaxiStefan** [reply]
I tested again on all 3 devices I mentioned above:

OnePlus (Android 13) did not have a problem using WAL
Samsung Galaxy Active Tab 3 (Android 12) and Samsung S8+ (Android 9) both crashed with the transaction exception

**Reply by MaxiStefan** [reply]
Ran these tests in debug as well as in release mode and no changes

**Reply by simolus3** [reply, marked as answer]
Oh no, I think the error is actually easy to fix and I've overlooked it the entire time. Error code 6410 means that the VFS was unable to store data in the temp path. On Android, that is entirely expected since applications aren't supposed to use `/tmp`. This would happen for every query requiring more memory than sqlite3 would like to allocate internally, I think WAL just makes this more likely.

This can likely be fixed by fixing the path `sqlite3` uses for temporary files (doing https://github.com/simolus3/drift/issues/876#issuecomment-710013503 in the `isolateSetup` parameter of `NativeDatabase.createInBackground` should do it. I'm sorry for not catching this earlier.

**Reply by MaxiStefan** [reply]
Yup, this seems to have fixed the issue. Thank you so much!
